File: card-api/src/main/java/com/jean202/cardmizer/api/normalization/MerchantNormalizationRule.java
package com.jean202.cardmizer.api.normalization;

import java.util.Objects;
import java.util.Set;

public record MerchantNormalizationRule(String category, Set<String> keywords, Set<String> inferredTags) {
    public MerchantNormalizationRule {
        if (category == null || category.isBlank()) {
            throw new IllegalArgumentException("category must not be blank");
        }
        keywords = Set.copyOf(Objects.requireNonNull(keywords, "keywords must not be null"));
        inferredTags = Set.copyOf(Objects.requireNonNull(inferredTags, "inferredTags must not be null"));
        if (keywords.isEmpty()) {
            throw new IllegalArgumentException("keywords must not be empty");
        }
    }
}

File: card-api/src/main/kotlin/com/jean202/cardmizer/api/normalization/MerchantNormalizationRule.kt
package com.jean202.cardmizer.api.normalization

class MerchantNormalizationRule(
    val category: String,
    keywords: Set<String>,
    inferredTags: Set<String>,
) {
    val keywords: Set<String>
    val inferredTags: Set<String>

    init {
        require(category.isNotBlank()) { "category must not be blank" }
        require(keywords.isNotEmpty()) { "keywords must not be empty" }
        this.keywords = keywords.toSet()
        this.inferredTags = inferredTags.toSet()
    }
}
